gateway/src/shard.rs

```diff
@@ -0,0 +1,968 @@
+//! Primary logic and implementation details of Discord gateway websocket
+//! connections.
+//!
+//! Shards are, at their heart, a websocket connection with some state for
+//! maintaining an identified session with the Discord gateway. For more
+//! information about what a shard is in the context of Discord's gateway API,
+//! refer to the documentation for [`Shard`].
+//!
+//! # Implementation Flow
+//!
+//! Other than informative methods and getters, shards have two functions:
+//! sending websocket messages and receiving websocket messages, called
+//! "events" in the context of the Discord gateway.
+//!
+//! Sending a message is simple and the flow of it looks like:
+//!
+//! 1. If the user sends a [command] via [`Shard::command`] the command is
+//! serialized into a raw websocket message and [`Shard::send`] is called;
+//! 2. The user calls [`Shard::send`] and the sending of the message goes
```

Comment:
step 1 and 2 have some overlap which makes it a bit strange to read.

Reply:
What would you suggest that continues to make it clear that Shard::command calls Shard::send?

Reply:
Maybe mark step one as optional and amf just say 2. When Shard::send is called...

Though tbh this is a pretty nitpicky thing

Reply:
Made an attempt at this.